engine/src/main/java/org/pentaho/di/job/Job.java

```diff
@@ -0,0 +1,2367 @@
+    //
+    JobEntryLogTable jobEntryLogTable = jobMeta.getJobEntryLogTable();
+    if ( jobEntryLogTable.isDefined() ) {
+      addJobListener( new JobAdapter() {
+        @Override public void jobFinished( Job job ) throws KettleException {
+          try {
+            writeJobEntryLogInformation();
+          } catch ( KettleException e ) {
+            throw new KettleException( BaseMessages.getString( PKG,
+                "Job.Exception.UnableToPerformJobEntryLoggingAtJobEnd" ), e );
+          }
+        }
+      } );
+    }
+
+    // If we need to write the log channel hierarchy and lineage information,
+    // add a listener for that too...
+    //
+    ChannelLogTable channelLogTable = jobMeta.getChannelLogTable();
+    if ( channelLogTable.isDefined() ) {
+      addJobListener( new JobAdapter() {
+
+        @Override public void jobFinished( Job job ) throws KettleException {
+          try {
+            writeLogChannelInformation();
+          } catch ( KettleException e ) {
+            throw new KettleException( BaseMessages.getString( PKG, "Job.Exception.UnableToPerformLoggingAtTransEnd" ),
+                e );
+          }
+        }
+      } );
+    }
+
+    JobExecutionExtension extension = new JobExecutionExtension( this, result, null, false );
+    ExtensionPointHandler.callExtensionPoint( log, KettleExtensionPoint.JobBeginProcessing.id, extension );
+
+    return true;
+  }
+
+  //
+  // Handle logging at end
+  /**
+   * End processing.
+   *
+   * @return true, if successful
+   * @throws KettleJobException
+   *           the kettle job exception
+   */
+  private boolean endProcessing() throws KettleJobException {
+    LogStatus status;
+    if ( !isActive() ) {
+      if ( isStopped() ) {
+        status = LogStatus.STOP;
+      } else {
+        status = LogStatus.END;
+      }
+    } else {
+      status = LogStatus.RUNNING;
+    }
+    try {
+      if ( errors.get() == 0 && result != null && !result.getResult() ) {
+        errors.incrementAndGet();
+      }
+
+      logDate = new Date();
+
+      /*
+       * Sums errors, read, written, etc.
+       */
+
+      JobLogTable jobLogTable = jobMeta.getJobLogTable();
+      if ( jobLogTable.isDefined() ) {
+
+        writeLogTableInformation( jobLogTable, status );
+      }
+
+      return true;
+    } catch ( Exception e ) {
+      throw new KettleJobException( e ); // In case something else goes wrong.
+    }
+  }
+
+  /**
+   * Writes information to Job Log table. Cleans old records, in case job is finished.
+   */
+  protected void writeLogTableInformation( JobLogTable jobLogTable, LogStatus status ) throws KettleJobException,
+    KettleDatabaseException {
+    boolean cleanLogRecords = status.equals( LogStatus.END );
+    String tableName = jobLogTable.getActualTableName();
+    DatabaseMeta logcon = jobLogTable.getDatabaseMeta();
+
+    Database ldb = createDataBase( logcon );
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
Analysis of this package determined that this line of code contains a resource that might not have closed properly. A resource leak can slow down or crash your system.

**Fix**
Consider closing the following resource in a try-finally block: *ldb*. The resource is referenced at lines 1179, 1181, 1182, 1183, 1186, 1193, 1194. The resource is closed at line 1196. There are other execution paths that do not contain closure statements, for example, when Database.commitLog throws an exception, and an if condition at line 1193 evaluates to true.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).